core/src/main/java/com/linecorp/armeria/server/AbstractServiceBindingBuilder.java

```diff
@@ -136,11 +137,21 @@ final void build0(HttpService service) {
                              ImmutableSet.of());
 
         final List<Route> routes = buildRouteList(fallbackRoutes);
+        final HttpService decoratedService = defaultServiceConfigSetters.decorator().apply(service);
         for (Route route : routes) {
-            final HttpService decoratedService = defaultServiceConfigSetters.decorator().apply(service);
             final ServiceConfigBuilder serviceConfigBuilder =
                     defaultServiceConfigSetters.toServiceConfigBuilder(route, decoratedService);
             serviceConfigBuilder(serviceConfigBuilder);
         }
     }
+
+    final void build0(HttpService service, Route mappedRoute) {
+        final List<Route> routes = buildRouteList(ImmutableSet.of());
+        checkState(routes.size() == 1, "routes.size() = %s (expected: 1)");
```

Comment:
Use `assert` if this can't happen. Otherwise please rephrase the exception message for better user experience.